### PR TITLE
[DC-1225] Add scripts directory and move files

### PR DESCRIPTION
**Pull request by s-rubenstein**
__Jira ticket__: https://broadworkbench.atlassian.net/browse/dc-1225

## Addresses
Aligns running TDR with best practices based on the [java template repo](https://github.com/DataBiosphere/terra-java-project-template). Enables running both locally and in docker. Moves db init scripts under the scripts tab. Updates render_configs.sh path and adds a `-e` flag which outputs the env variables. This is useful for `eval`ing the script in order to set your env variables. I also updated the getting started doc and removed the database readme, as the scripts should handle all DB setup for you.

Tests fail as the actions PR needs to merge first.

### CODE REVIEW

File: scripts/run [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+- local           run as a local terminal process
+- docker          run in a docker container
+EOF
+}
+
+# script variables and defaults
+source $(dirname $0)/init.sh
+source "${SCRIPTS_DIR}"/render_configs.sh
+
+JADE_USER_EMAIL=<PUT EMAIL HERE>
```
Comment (s-rubenstein) [inline scripts/run]: I think the trick is to put the environment variables here? But it doesn't actually pick it up. (It seems like?)

I also can get it to run bootRun here, but not with docker, so that still needs to get fixed.

File: .travis.yml [hunk 2]
```diff
@@ -30,7 +30,7 @@ before_install:
   - source $HOME/google-cloud-sdk/path.bash.inc
   - gcloud config set project $INTEGRATION_PROJECT
   - gcloud config set compute/zone $INTEGRATION_ZONE
-  - ./render-configs.sh $TRAVIS_TOKEN
+  - ./scripts/render_configs.sh $TRAVIS_TOKEN
```
Comment (pshapiro4broad) [inline .travis.yml]: I don't think we use travis anymore, I would delete this file rather than update it

File: DATABASE.md [hunk 3]
```diff
@@ -1,56 +0,0 @@
-### Install
```
Comment (s-rubenstein) [inline DATABASE.md:1]: We no longer need this because the scripts should take care of it for us - and docker will manage versions.

**Comment by fboulnois** [comment]
Consider running `shellcheck` on all the scripts, I see a few failures.

**Comment by s-rubenstein** [comment]
@fboulnois good call, I had sorta thought the existing code copied from the template repo would be okay, but looks like there might be a few shellcheck things to do over there. I might PR those there too.

File: docs/jade-getting-started.md [hunk 4]
```diff
@@ -377,24 +328,24 @@ Instead, you can select a specific test to run either in Intellij or the command
 First, make sure you have run through the following steps:
 * Start postgres
 * Ensure docker is running
-* You may need to re-auth with vault every so often. Run `vault login -method=github token=$(cat ~/.gh_token)`
-* Run `./render-configs.sh -a integration` to pull secrets from vault. For Azure Integration tests,
+* Auth as your broadinstitute.org to pull from Google Secrets Manager
```
Comment (rjohanek) [inline docs/jade-getting-started.md]: the old instructions say how to auth with vault maybe we should now include the auth command for google here and other places it says this

File: docs/jade-getting-started.md [hunk 5]
```diff
@@ -377,24 +328,24 @@ Instead, you can select a specific test to run either in Intellij or the command
...
+* Run `./scripts/render_configs.sh -a integration` to pull secrets from vault. For Azure Integration tests,
```
Comment (rjohanek) [inline docs/jade-getting-started.md]: is it still to pull secrets from vault?

File: docs/jade-getting-started.md [hunk 6]
```diff
@@ -301,26 +272,8 @@ export JADE_USER_EMAIL=<EMAIL_YOU_CREATED_FOR_DEVELOPMENT>
...
-export AZURE_SYNAPSE_SQLADMINUSER=$(cat /tmp/jade-dev-synapse-admin-user.key)
-export AZURE_SYNAPSE_SQLADMINPASSWORD=$(cat /tmp/jade-dev-synapse-admin-password.key)
-export AZURE_SYNAPSE_ENCRIPTIONKEY=$(cat /tmp/jade-dev-synapse-encryption-key.key)
 export AZURE_SYNAPSE_INITIALIZE=false
 
-# RBS
-# Defaults to RBS tools; you can switch to dev by running `./render-configs.sh -r dev`
-export RBS_POOLID=$(cat /tmp/rbs-pool-id.txt)
-export RBS_INSTANCEURL=$(cat /tmp/rbs-instance-url.txt)
-
```
Comment (rjohanek) [inline docs/jade-getting-started.md:323]: are the deleted env variables now automatically populated by running render configs?

File: docs/jade-getting-started.md [hunk 7]
```diff
@@ -377,24 +328,24 @@ Instead, you can select a specific test to run either in Intellij or the command
...
 * Run `./gradlew :testIntegration --tests '*<test name>'` to run a specific integration test (e.g `./gradlew :testIntegration --tests '*testSnapshotBuilder'`)
 
 ** Run or Debug test in Intellij **
 * If you just refreshed your Z-shell configuration, you may need to restart intellij to get the
   environment variables to populate the Intellij run configurations. Alternatively, you can run
-  `./render-configs.sh -i -a integration` which will put all the environment variables into your clipboard and then you
+  `./scripts/render_configs.sh -i -a integration` which will put all the environment variables into your clipboard and then you
   can paste them into the Intellij test setup.
-* Start application by running `./gradlew bootRun`
+* Start application by running `./scripts/run local` (or in docker)
```
Comment (rjohanek) [inline docs/jade-getting-started.md]: maybe you could include the docker command here too? probably fine either way

File: docs/jade-getting-started.md [hunk 8]
```diff
@@ -454,8 +405,10 @@ Ensure that:
     ```
     vault login -method=github token=$(cat ~/.gh_token)
```
Comment (rjohanek) [inline docs/jade-getting-started.md]: do we still need this?

File: docs/jade-getting-started.md (hunk 4 above)
Comment (s-rubenstein) [reply docs/jade-getting-started.md]: Ah, sure, can do.

File: docs/jade-getting-started.md [hunk 9]
```diff
@@ -271,24 +247,19 @@ Create a "tdr-dev" managed application:
...
+values. If you are using the setup script, `./scripts/render_configs.sh` should automatically run.
+An alternate is to run `./scripts/render_configs.sh -i` which will put the variables into your clipboard. You
 can then paste these values into an intellij bootRun or test run profile.
 
 ### Environment Variables
 
 While not exhaustive, here's a list that notes the important environment variables to set when running
-`jade-data-repo` locally. These variables override settings in jade-data-repo/application.properties.
+`jade-data-repo` locally that are not set by `./scripts/render_configs.sh`. These variables override settings in jade-data-repo/application.properties.
 You can convert any application.property to an environment variable by switching to upper case and
```
Comment (rjohanek) [inline docs/jade-getting-started.md:229]: besides the user email, why can't the other env vars be populated by render configs?

Comment (s-rubenstein) [reply docs/jade-getting-started.md:229]: Some of them probably could be - I wanted to hold off on making too many changes to render_configs as part of this, other than what was needed to get the scripts up and running. I would probably advocate for looking into that being a separate ticket. If you'd like, I can file that.

File: docs/jade-getting-started.md [hunk 10]
```diff
@@ -271,24 +247,19 @@ Create a "tdr-dev" managed application:
...
 every "." to "_".
```
Comment (rjohanek) [inline docs/jade-getting-started.md:231]: following line references ZZ but this is not used in this step. could be for a later clean up ticket though.

File: DATABASE.md [hunk 11]
```diff
@@ -1,56 +0,0 @@
...
-5. Start the server
-
-## Setting up the database
-You'll want to follow all the steps below and stop once you have updated to the latest migration.
-
-### Create the data repo db and user
-
-    psql -f db/create-data-repo-db
-
-### Clear the database
```
Comment (pshapiro4broad) [inline DATABASE.md:21]: Some info in here is no longer needed, but for other info I don't see where it's documented or available in the new code. For example, `truncate-tables` doesn't seem to exist in the new code, nor does any of the liquibase operations below.

File: docs/jade-getting-started.md (hunk 10 above)
Comment (s-rubenstein) [reply docs/jade-getting-started.md:231]: Yeah, I'd say lets follow up on that, when we get to a more comprehensive readme update

File: scripts/run [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+                  either `local` or `docker`
+- local           run as a local terminal process
+- docker          run in a docker container
+- check           run linter and checks
+EOF
+}
+
+# script variables and defaults
+source "$(dirname "$0")"/init.sh
+eval "$("${SCRIPTS_DIR}"/render_configs.sh -e)"
```
Comment (pshapiro4broad) [inline scripts/run]: This causes `render_configs.sh` to run even if a user is trying to view the help. Maybe this could go into a function and only be called if necessary? With vault, `render_configs` was pretty fast, but gcloud seems much slower.

File: scripts/render_configs.sh
Comment (pshapiro4broad) [inline scripts/render_configs.sh:1]: Why was this renamed? We do have a mix of `-` and `_` in our script names and I don't see a common convention yet. Personally I prefer `-` for names as it's easier (for me) to type.

File: scripts/render_configs.sh [hunk 13]
```diff
@@ -68,12 +72,18 @@ if $USE_VAULT; then
 else
   AZURE_SECRETS=$(gcloud secrets versions access latest --project $GCLOUD_PROJECT --secret azure-secrets)
 fi
-
+# These are used in the variable echoing or setting to clipboard below
+# shellcheck disable=SC2034
 AZURE_CREDENTIALS_HOMETENANTID=$(echo "$AZURE_SECRETS" | jq -r '."tenant-id"' | tee /tmp/jade-dev-tenant-id.key)
```
Comment (pshapiro4broad) [inline scripts/render_configs.sh]: Is there still something using the tee'd output files? I don't see it used anymore.

File: scripts/render_configs.sh [hunk 14]
```diff
@@ -130,25 +142,35 @@ else
...
     SETTINGS+="$VAR_NAME=${!VAR_NAME};"
   done
 
   # Copy variables to clipboard
+  echo "# Environment variables copied to clipboard"
   echo "$SETTINGS"  | pbcopy
-  echo "Environment variables copied to clipboard"
+fi
+
+if [[ "${ECHO_SETTINGS_TO_OUTPUT}" == "y" ]]; then
```
Comment (pshapiro4broad) [inline scripts/render_configs.sh]: Quotes aren't needed around `y` here. Also if we use `true` and `false`, then you can simplify the `if`
```suggestion
ECHO_SETTINGS_TO_OUTPUT=false

...
   ECHO_SETTINGS_TO_OUTPUT=true
...

if "${ECHO_SETTINGS_TO_OUTPUT}"; then
```

File: DATABASE.md (hunk 11 above)
Comment (s-rubenstein) [reply DATABASE.md:21]: Yeah, `truncate-tables` I moved over from the old DB folder. I wasn't sure what to do with it. Definitely unneeded

File: scripts/render_configs.sh (hunk 13 above)
Comment (s-rubenstein) [reply scripts/render_configs.sh]: I don't know, but I am hesitant to change render-configs now as part of this PR, beyond what is needed to bring in the scripts.

File: docs/jade-getting-started.md (hunk 6 above)
Comment (s-rubenstein) [reply docs/jade-getting-started.md:323]: Yes

File: DATABASE.md [hunk 15]
```diff
@@ -1,56 +0,0 @@
...
-
-    psql -f db/truncate-tables datarepo drmanager
-
-### Migrations
```
Comment (pshapiro4broad) [inline DATABASE.md:25]: This section is still relevant and not covered elsewhere. I would either restore this part of the readme (or move it to another readme) or move it to a script, maybe `setup`.

File: build.gradle [hunk 16]
```diff
@@ -27,7 +27,7 @@ buildscript {
 }
 
 plugins {
-    id 'com.google.cloud.tools.jib' version '3.4.3'
+    id 'com.google.cloud.tools.jib' version '3.4.2'
```
Comment (pshapiro4broad) [inline build.gradle:31]: Maybe add a comment here about why this isn't using the current version

File: docs/jade-getting-started.md [hunk 17]
```diff
@@ -115,8 +115,6 @@ upgrading Kubernetes deployments, which are otherwise challenging to manage.
 Some manual configuration is required below.
 5. [Helmfile](https://github.com/roboll/helmfile) streamlines deploying multiple
 helm charts.
-6. [Vault](https://www.vaultproject.io/) is an encrypted database used to store
```
Comment (pshapiro4broad) [inline docs/jade-getting-started.md:118]: Markdown will renumber these automatically, but it seems odd to leave them missing an entry in the source. I would either renumber the entries after this, or change all entries to be `1.` so it's clear that markdown is going to handle the numbering.

File: docs/jade-getting-started.md [hunk 18]
```diff
@@ -115,8 +115,6 @@ upgrading Kubernetes deployments, which are otherwise challenging to manage.
 Some manual configuration is required below.
```
Comment (pshapiro4broad) [inline docs/jade-getting-started.md:112]: On the line above, `Docker` is mentioned; @rjohanek ran into problems using the brew cask installed docker. We were able to fix them by uninstalling brew's docker and then using the broad installed docker. Since running these scripts now depend on docker, it would be good to clean up the docs around setting it up as part of this. To do this, I would remove the references to using brew to install docker here, and from the `Brewfile`.

File: scripts/build [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+source "$(dirname "$0")"/init.sh
+skip_tests=""
+
+
+# process command-line options (if any)
+die() { log_error "$*" >&2; echo ""; usage; exit 2; }  # complain to STDERR and exit with error
+needs_arg() { if [ -z "$OPTARG" ]; then die "No arg for --$OPT option"; fi; }
+
+while getopts ht-: OPT; do  # allow -h and -- "with arg"
+  # support long options: https://stackoverflow.com/a/28466267/519360
```
Comment (pshapiro4broad) [inline scripts/build:27]: I don't think the extra work to support long options is worth it, and I think we'd be better off only using short options. It adds cruft to the code without much benefit IMO.

File: scripts/compose.yaml [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+name: terradatarepo
+
+services:
+
+  service:
+    image: jade-data-repo
+    depends_on:
+      - database
+    environment:
+      - DATABASE_NAME=database
```
Comment (pshapiro4broad) [inline scripts/compose.yaml]: Why is this needed? I don't think TDR is using this env var.

Maybe it's used by application.properties? If so, I wonder why the value would be `database`?

File: scripts/render-configs.sh [hunk 21]
```diff
@@ -130,25 +142,35 @@ else
...
+  # Loop over the array
+  for VAR_NAME in "${VARIABLE_NAMES[@]}"; do
+    # Append the variable name and its value to the string
+    OUTPUT+="export $VAR_NAME='${!VAR_NAME}';"
+  done
+
+  echo "$OUTPUT"
 fi
 
 unset AZURE_ENV
```
Comment (pshapiro4broad) [inline scripts/render-configs.sh:172]: This is only needed if callers use `source`. Once everyone is using `eval` with `-e` then we no longer have to worry about exporting extra variables

File: scripts/run [hunk 22]
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env bash
+# run the service in one of different deployment modes
+set -e #-x
+
+usage() {
+  cat <<-'EOF'
+usage: run [-h|--help] RUNTYPE
+[-h|--help]       print this help text
+RUNTYPE is one of:
+- tests           run tests against repo
```
Comment (pshapiro4broad) [inline scripts/run]:
```suggestion
- tests           run unit tests
```

File: scripts/run [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+- integration     run integration tests against running instance;
+                  either `local` or `docker`
```
Comment (pshapiro4broad) [inline scripts/run]: Right now `run_integration_tests()` always uses docker and doesn't support `local`.
```suggestion
- integration     run integration tests against a docker instance
```

File: scripts/run [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+- local           run as a local terminal process
+- docker          run in a docker container
+- check           run linter and checks
```
Comment (pshapiro4broad) [inline scripts/run]: I don't think this is accurate; I believe `gradle check` will run all unit tests too. AFAIK we don't have a lint type phase. Maybe running sonar/srcclr is the equivalent to a linter? But running sonar and srcclr requires a token, which this script doesn't have (yet)

File: scripts/run [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+  # this is run via both:
+  # - ./github/workflows/build-and-test.yml
+  # - ./github/workflows/integration-tests.yml
+  #
+  # please see integration-tests.yml for more information re: inputs.environment and TEST_ENV
+  TEST_ENV=${TEST_ENV:-local}
+
+  cd "${ROOT_DIR}"
+  export IT_JADE_API_URL=http://localhost:8080
+  AZURE_ENV=integration
```
Comment (pshapiro4broad) [inline scripts/run]: It's not clear but setting this env var changes the way `run_docker` works. It would be better to make `AZURE_ENV` a parameter to `run_docker` to make this relationship clearer.

File: scripts/run [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+  run_docker
+  # shellcheck disable=SC2086 # The array syntax and quoting seems to break gradle
+  ./gradlew ${GRADLE_OPTS} testIntegration
+
+  # for developer runs, build an integration report locally
+  if [ -z "$CI" ]; then
+    log_debug "TODO: how do we build a report?"
+  fi
```
Comment (pshapiro4broad) [inline scripts/run:71]: I would remove this from here and below; the junit phase creates a test report automatically and prints the location of the report in the build output.

File: scripts/run [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+#./gradlew --build-cache srcclr
+}
+
+run_integration_tests() {
+  # NOTE: interesting - no integration-tests.yml have been triggered or run
+  # this is run via both:
+  # - ./github/workflows/build-and-test.yml
+  # - ./github/workflows/integration-tests.yml
+  #
+  # please see integration-tests.yml for more information re: inputs.environment and TEST_ENV
```
Comment (pshapiro4broad) [inline scripts/run]: Is this comment still relevant?

File: scripts/run [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+  TEST_ENV=${TEST_ENV:-local}
```
Comment (pshapiro4broad) [inline scripts/run:61]: As far as I can tell, `TEST_ENV` is not used anywhere and can be removed.

File: scripts/run [hunk 29]
```diff
@@ -0,0 +1,127 @@
...
+
+  cd "${ROOT_DIR}"
+  export IT_JADE_API_URL=http://localhost:8080
```
Comment (pshapiro4broad) [inline scripts/run]: This is not needed for connected tests

File: scripts/run [hunk 30]
```diff
@@ -0,0 +1,127 @@
...
+  run_docker
```
Comment (pshapiro4broad) [inline scripts/run]: This is also not needed for connected tests; connected tests are run like unit tests (they run against code running in the same java space as the test itself) but unlike unit tests, they connect to live cloud services.

File: scripts/run [hunk 31]
```diff
@@ -0,0 +1,127 @@
...
+  # NOTE: interesting - no integration-tests.yml have been triggered or run
+  # this is run via both:
+  # - ./github/workflows/build-and-test.yml
+  # - ./github/workflows/integration-tests.yml
+  #
+  # please see integration-tests.yml for more information re: inputs.environment and TEST_ENV
+  TEST_ENV=${TEST_ENV:-local}
+
+  cd "${ROOT_DIR}"
+  export IT_JADE_API_URL=http://localhost:8080
```
Comment (pshapiro4broad) [inline scripts/run]: I would move this so it's just above `./gradlew` since it's not needed for `run_docker`

File: scripts/run [hunk 32]
```diff
@@ -0,0 +1,127 @@
...
+  esac
+done
+shift $((OPTIND-1)) # remove parsed options and args from $@ list
+
+
+# process positional arguments (if any)
+RUNTYPE=${1}
+
+run_tests() {
+  eval "$("${SCRIPTS_DIR}"/render-configs.sh -a "$AZURE_ENV" -e)"
```
Comment (pshapiro4broad) [inline scripts/run]: I don't think unit tests need `render_configs.sh` to run. In fact, I don't think they need any env vars set at all.

File: scripts/run [hunk 33]
```diff
@@ -0,0 +1,127 @@
...
+  # shellcheck disable=SC2086 # The array syntax and quoting seems to break gradle
+  ./gradlew ${GRADLE_OPTS} testConnected
+
+  # for developer runs, build an integration report locally
+  if [ -z "$CI" ]; then
+    log_debug "TODO: how do we build a report?"
+  fi
+}
+
+run_local() {
```
Comment (pshapiro4broad) [inline scripts/run]: This script uses `run` to mean two different things. In this case, `run_local` starts a running copy of the service. Whereas `run_test` runs the unit tests. Maybe this should be `start_local` and `start_docker`?

File: scripts/run [hunk 34]
```diff
@@ -0,0 +1,127 @@
...
+  eval "$("${SCRIPTS_DIR}"/render-configs.sh -a "$AZURE_ENV" -e)"
+  cd "$ROOT_DIR"
+  # shellcheck disable=SC2086 # The array syntax and quoting seems to break gradle
+  ./gradlew ${GRADLE_OPTS} check
+}
+
+case "$RUNTYPE" in
+  tests )        run_tests ;;
+  integration )  run_integration_tests ;;
+  connected )    run_connected_tests ;;
```
Comment (pshapiro4broad) [inline scripts/run]: Need to add a usage doc for `connected` run type

File: scripts/run [hunk 35]
```diff
@@ -0,0 +1,127 @@
...
+
+run_check() {
+  eval "$("${SCRIPTS_DIR}"/render-configs.sh -a "$AZURE_ENV" -e)"
+  cd "$ROOT_DIR"
+  # shellcheck disable=SC2086 # The array syntax and quoting seems to break gradle
+  ./gradlew ${GRADLE_OPTS} check
+}
+
+case "$RUNTYPE" in
+  tests )        run_tests ;;
```
Comment (pshapiro4broad) [inline scripts/run]: We don't have a shell formatting linter, but most of our existing bash code omits the space before the `)` and the `;;`. Typically the case body is formatted across lines too but IMO one line is fine too. But I would make the spacing agree with the existing code
```suggestion
  tests)        run_tests;;
```

File: src/main/resources/application.properties [hunk 36]
```diff
@@ -13,13 +13,13 @@ spring.web.resources.cache.cachecontrol.must-revalidate=true
 spring.web.resources.static-locations=classpath:/api/
 db.migrate.dropAllOnStart=false
 db.migrate.updateAllOnStart=true
-db.datarepo.uri=jdbc:postgresql://127.0.0.1:5432/datarepo
+db.datarepo.uri=jdbc:postgresql://${DATABASE_NAME:127.0.0.1}:5432/datarepo
```
Comment (pshapiro4broad) [inline src/main/resources/application.properties]: Why is `DATABASE_NAME` needed here? Is that due to running inside docker? If so, does this still work when running outside docker?

**Review by pshapiro4broad — COMMENTED** [review]
A few more comments / suggestions

File: src/main/resources/application.properties (hunk 36 above)
Comment (s-rubenstein) [reply src/main/resources/application.properties]: Due to running inside docker. The second bit is a default, so if DATABASE_NAME isn't set, it defaults to the same value

File: build.gradle (hunk 16 above)
Comment (s-rubenstein) [reply build.gradle:31]: Would you mind copying the article you found about the latest version being broken?

File: docs/jade-getting-started.md (hunk 9 above)
Comment (rjohanek) [reply docs/jade-getting-started.md:229]: yeah sounds good thanks

File: docs/jade-getting-started.md [hunk 37]
```diff
@@ -333,68 +255,54 @@ export GOOGLE_ALLOWREUSEEXISTINGBUCKETS=true
...
+* Run `./scripts/run-db start` to start the DB in a docker container
+* Run `./scripts/run local` to run TDR locally or `./scripts/run docker` to run TDR in a docker container
+* To Build the code and run the unit tests:
 
 ```
-./gradlew build           # build jade-data-repo and run unit tests
-./gradlew bootRun         # build jade-data-repo with Spring Boot features
-./gradlew check           # linters and unit tests
+./scripts/build project   # build jade-data-repo and run unit tests
+./scripts/run check       # linters and unit tests
```
Comment (rjohanek) [inline docs/jade-getting-started.md]: what is the difference between project and check? because doesn't the project need to be built in order to run tests and don't linters run when the tests run?

File: scripts/render-configs.sh (hunk 21 above)
Comment (s-rubenstein) [reply scripts/render-configs.sh:172]: I think people might still use `source`, especially while we're getting used to the new model. My inclination is to leave it and clean up later if needed.

File: scripts/build (hunk 19 above)
Comment (s-rubenstein) [reply scripts/build:27]: @snf2ye did mention to me she prefers the long options, so given that I think it seems fine to add. I think we could eventually align on one option eventually.

**Review by snf2ye — APPROVED** [review]
🎉

**Review by rjohanek — APPROVED** [review]
